### PR TITLE
On Android, attach the threadpool threads to the JavaVM

### PR DESCRIPTION
**Pull request by lubeltra**
libHttpClient needs to be pumped on a thread that is attached to the JavaVM to function on Android.
When using `XTaskQueueDispatchMode::ThreadPool` the threads the pumping is done are managed by the task system, so we need a way to attach those to the JavaVM.

This change adds a function `XTaskQueueSetJvm` that takes a pointer to the JavaVM, the threadpool threads will detect it and use it from then on. `HCInitialize` will also implicitly do this when called.

### CODE REVIEW

File: Build/libHttpClient.141.Android.C/libHttpClient.141.Android.C.vcxproj [hunk 1]
```diff
@@ -131,8 +131,9 @@
     <ClInclude Include="$(MSBuildThisFileDirectory)..\..\Include\XAsync.h" />
     <ClInclude Include="$(MSBuildThisFileDirectory)..\..\Include\XAsyncProvider.h" />
     <ClInclude Include="$(MSBuildThisFileDirectory)..\..\Include\XTaskQueue.h" />
+    <ClInclude Include="..\..\Include\httpClient\async_jvm.h" />
```
Comment (matt-vanderkolk) [inline Build/libHttpClient.141.Android.C/libHttpClient.141.Android.C.vcxproj]: >="..\..\ [](start = 22, length = 8)

MSBuildThisFileDirectory?

File: Build/libHttpClient.141.Android.C/libHttpClient.141.Android.C.vcxproj.filters [hunk 2]
```diff
@@ -186,6 +186,9 @@
     <ClInclude Include="$(MSBuildThisFileDirectory)..\..\Include\XTaskQueue.h">
       <Filter>C++ Public Includes</Filter>
     </ClInclude>
+    <ClInclude Include="..\..\Include\httpClient\async_jvm.h">
```
Comment (matt-vanderkolk) [inline Build/libHttpClient.141.Android.C/libHttpClient.141.Android.C.vcxproj.filters]: >="..\..\ [](start = 22, length = 8)

MSBuildThisFileDirectory

File: Build/libHttpClient.141.Android.C/libHttpClient.141.Android.C.vcxproj (hunk 1 above)
Comment (matt-vanderkolk) [reply Build/libHttpClient.141.Android.C/libHttpClient.141.Android.C.vcxproj]: Should you be updating the CMake stuff?

---
In reply to: [262319532](https://github.com/Microsoft/libHttpClient/pull/387#discussion_r262319532) [](ancestors = 262319532)

File: Include/httpClient/async_jvm.h [hunk 3]
```diff
@@ -0,0 +1,9 @@
+#pragma once
```
Comment (matt-vanderkolk) [inline Include/httpClient/async_jvm.h]: >pragma [](start = 1, length = 6)

copyright header

File: Source/Task/ThreadPool_stl.cpp [hunk 4]
```diff
@@ -43,6 +47,11 @@ class ThreadPoolImpl
                 numThreads--;
                 m_pool.emplace_back(std::thread([this]
                 {
+#if defined(HC_PLATFORM) && HC_PLATFORM == HC_PLATFORM_ANDROID
+                    JNIEnv* jniEnv = nullptr;
+                    JavaVM* jvm = s_javaVm;
```
Comment (matt-vanderkolk) [inline Source/Task/ThreadPool_stl.cpp]: >s_javaVm [](start = 34, length = 8)

it feels like you should be initializing this to null here

Comment (lubeltra) [reply Source/Task/ThreadPool_stl.cpp]: yes, remainder from a previous iteration where I only read `s_javaVm` once

File: Build/libHttpClient.141.Android.C/libHttpClient.141.Android.C.vcxproj (hunk 1 above)
Comment (lubeltra) [reply Build/libHttpClient.141.Android.C/libHttpClient.141.Android.C.vcxproj]: yes

**Pull request merged**